Concept: Arithmetic (#726)
exercism/bash: concepts/arithmetic/introduction.md

# Introduction to Arithmetic in the Bash Shell

Bash can do arithmetic but _only with integers_.
If you need to do floating-point math, you'll need to call out to some external program (like [awk][awk] or [bc][bc]).

The arithmetic operators are listed (in order of precedence) in the manual: [Shell Arithmetic][arithmetic].

Additionally, Bash only performs arithmetic in certain places (known as arithmetic "contexts").

## Arithmetic Expansion

> Arithmetic expansion looks like parameter expansion, but with double parentheses.

Reviewer: Did you already introduce command substitution? Arithmetic expansion looks more like that. I wouldn't really think arithmetic expansion looks like PE a ton.

```bash

> $((expression))

Reviewer: Were we using spaces inside `(( ... ))` in the last doc?

Author: I'll make a note that spaces are allowed.

Reviewer: I meant more that we should use a consistent style across docs.

Author: Yes, in the Functions doc.

What's your preference? Spaces or no spaces

Reviewer: I prefer spaces but I'm not strongly opinionated here.

```

> This expansion returns the result of evaluating the expression.

Reviewer: ```suggestion
This expansion is expanded to the result of evaluating the expression.
```

> It is common to capture the result in a variable:

Reviewer: ```suggestion
It is commonly used to capture the result in a variable:
```

```bash
x=$((3 + 4))
```

## Arithmetic Conditional

The arithmetic conditional is performed within double parentheses _without_ the leading `$`:

```bash
((expression))
```

> This does not return the result.

Reviewer: ```suggestion
This is not replaced with the result.
```

The exit status is sometimes referred to as the return code. This does produce a result which sets the return code.

However it does produce an _exit status_.

* The exit status is 0 (success) if the expression's result is **non-zero**.
* The exit status is 1 (failure) if the expression's result is **zero**.

~~~~exercism/note

> Bash uses the same sense of true/false as the C language: zero is false, non-zero is true.

Reviewer: ```suggestion
Inside `((` Bash uses the same sense of true/false as the C language: zero is false, non-zero is true.
**This is the opposite of how Bash evaluates exit statuses!**
```

~~~~

This syntax is typically used in two places:

1. as the COMMAND part of an [`if` statement][conditional-if] or a [`while` loop][looping-while].

   ```bash
   if ((x == y - 1)); then ...
   while ((y > 0)); do ...
   ```

1. as a standalone command to modify a variable's value.

   ```bash

>    ((++x))

Reviewer: You should pre-increment here then discuss post-increments. This is slightly confusing.

   echo "new value for x is $x"
   ```

   ~~~~exercism/caution

>    Be careful using arithmetic this way combined with `set -e` that the expression does not result in zero.
>    The exit status will be non-zero.

Reviewer: ```suggestion
   Be careful using arithmetic this way combined with `set -e`.
   If the expression evaluates to zero, the exit status will be non-zero and Bash will terminate.
```

   ```bash
   x=0
   ((x++))
   echo "you will see this"

   set -e
   x=0
   ((x++))
   echo "shell exits on the previous line; you won't see this"
   ```
   ~~~~

## Other Places With Arithmetic Context

There are a few other places that are arithmetic contexts:

> * The [arithmetic for-loop][looping-for].

Reviewer: Should this be discussed/shown above?

Author: I don't think so. `for` requires 3 expressions separated by semicolons, which is specialized synax.

Reviewer: Sounds good. Having it in the looping doc instead seems reasonable, too.

> * The index part of a numerically-indexed array.

Reviewer: ```suggestion
* The index part of a (non-associative) array.
```

Author: I want to emphasis the "numeric" aspect.

Reviewer: If you make an associative array and use numbers as keys, is it numerically indexed?

Author: Nope.
```
x=5
declare -A a
a[x + 1]=10
declare -p a
# => declare -A a=(["x + 1"]="10" )

```

Author: But beyond the details, that's how I denote the 2 kinds of arrays. And that's the language I use in the Arrays doc.

Reviewer: ```bash
declare -A a
a[5]=10
a[6]=12
a[7]=14
declare -p a
# declare -A a=([7]="14" [6]="12" [5]="10" )
```

Aside from the unbalanced spacing, that looks like a numerically indexed array ;) I think the language/term can be ambiguous. But if you prefer that language, that works for me.

Author: This would be a good place for a `(For more info on arrays, see the [Arrays] concept doc)` link.

  ```bash
  echo "The next element is ${array[x + 1]}"
  # ................................^^^^^
  ```

* The offset and length parts of the `${var:offset:length}` expansion.
* The [`let` builtin command][let].
  * Note that `((expression))` is preferred over `let "expresson"`.
    Unlike the arithmetic conditional construct, the `let` expression is subject to filename-expansion and word-splitting, and it requires quoting.
* Variable assignment, when the variable has the "integer attribute".

  ```bash
  # variable x does not have the integer attribute
  x="3 + 4"
  declare -p x      # => declare -- x="3 + 4"

  # variable y has the integer attribute
  declare -i y
  y="4 + 5"
  declare -p y      # => declare -i y="9"
  ```

  ~~~~exercism/note
  It is the author's opinion that `declare -i` has limited usage.
  It's main benefit is to allow "bare" arithmetic (without the arithmetic expression syntax).
  This can cause confusion for readers of your code, where arithmetic expressions are evaluated in unexpected places.
  ~~~~

## Using Variables in an Arithmetic Expressions

> There is one aspect of Bash arithmetic that improves readability.

Reviewer: ```suggestion
Here is one aspect of arithmetic contexts that improves readability.
```

```bash
height=10
width=20
area=$((height * width))
```

> Did you see that the `$` dollar signs are missing from the variables in the arithmetic expression?

Reviewer: ```suggestion
Did you notice that the `$` dollar signs are missing from the variables in the arithmetic expression?
```

Bash lets you do that.
It makes expressions much easier to read.

> The manual says it this way:

Reviewer: ```suggestion
The manual says,
```
